sdk/core/Azure.Core/src/SyncCollection.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+
+namespace Azure
+{
+    public abstract class SyncCollection<T> : IEnumerable<T> where T : notnull
+    {
+        /// <summary>
+        /// Gets a <see cref="CancellationToken"/> used for requests made while
+        /// enumerating asynchronously.
+        /// </summary>
+        protected virtual CancellationToken CancellationToken { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncCollection{T}"/>
+        /// class for mocking.
+        /// </summary>
+        protected SyncCollection() =>
+            CancellationToken = CancellationToken.None;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncCollection{T}"/>
+        /// class.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken"/> used for requests made while
+        /// enumerating asynchronously.
+        /// </param>
+        protected SyncCollection(CancellationToken cancellationToken) =>
+            CancellationToken = cancellationToken;
+
+        /// <summary>
+        /// Enumerate the values a <see cref="Page{T}"/> at a time.  This may
+        /// make mutliple service requests.
+        /// </summary>
+        /// <param name="continuationToken">
+        /// A continuation token indicating where to resume paging or null to
+        /// begin paging from the beginning.
+        /// </param>
+        /// <param name="pageSizeHint">
+        /// The size of <see cref="Page{T}"/>s that should be requested (from
+        /// service operations that support it).
+        /// </param>
+        /// <returns>
+        /// An async sequence of <see cref="Page{T}"/>s.
+        /// </returns>
+        public abstract IEnumerable<Page<T>> ByPage(
```

Comment:
If this is the only additional declaration on an otherwise `IEnumerable<T>`, why not just an extension method? Keeping the inputs/outputs as least-derives makes testing/mocking easier. E.g. mocking a method that took `IEnumerable<T>` could just return an array.

Reply:
How would you implement such an extension method?

Reply:
If the `this` implementation is this class, you could access internals; otherwise, it wouldn't support pages. Guess it depends on expected behavior what you do then.

Reply:
Right, but then you need an extension method for every model that is returned as part of Enumerable and it would be impossible to mock it because it would be internal.